lib/local-cluster/test/unit/Control/Monitoring/MonitorSpec.hs

```diff
@@ -0,0 +1,89 @@
+module Control.Monitoring.MonitorSpec
+    ( spec
+    )
+where
+
+import Prelude
+
+import Control.Concurrent
+    ( threadDelay
+    )
+import Control.Foldl
+    ( Fold
+    )
+import Control.Monad
+    ( replicateM_
+    , (<=<)
+    )
+import Control.Monitoring.Folder
+    ( mkTracingFromFold
+    )
+import Control.Monitoring.Monitor
+    ( Monitor
+    , mkMonitor
+    , observe
+    , step
+    , trace
+    )
+import Control.Monitoring.Tracing
+    ( AnyTracing (AnyTracing)
+    , StateS (..)
+    )
+import Data.Foldable
+    ( forM_
+    )
+import Test.Hspec
+    ( Spec
+    , describe
+    , it
+    , shouldBe
+    )
+import UnliftIO
+    ( async
+    , forConcurrently_
+    , link
+    )
+
+import qualified Control.Foldl as L
+
+mkTracing :: StateS x -> Fold a b -> AnyTracing a b
+mkTracing w f = AnyTracing w $ mkTracingFromFold f w
+
+test
+    :: (Show b, Eq b)
+    => Int
+    -> StateS x
+    -> Fold a b
+    -> b
+    -> (Int -> Monitor IO a b -> IO ())
+    -> IO ()
+test cycles state fold target action = forM_ [1 .. cycles] $ \cycle' -> do
+    monitor <- mkMonitor (mkTracing state fold) pure
```

Comment:
Note: The purpose of running multiple cycles is to tease out race conditions. A systematic way to do that would be [Control.Monad.IOSimPOR](https://github.com/input-output-hk/io-sim/blob/main/io-sim/how-to-use-IOSimPOR.md) — but that's probably overkill here; I just wanted to mention it.